tests/test_core.py

```diff
@@ -634,3 +634,13 @@ def test_array_creation_with_invalid_fields():
 def test_promote_nullable():
     with pytest.warns(DeprecationWarning):
         assert ndx.promote_nullable(np.int64) == ndx.nint64
+
+
+@pytest.mark.parametrize(
+    "val", [1, 1.0, 1.5, 0.123456789, "a", np.float64(0.123456789)]
+)
+def test_scalar_promote(val):
+    x = ndx.asarray([val] * 4)
+    actual = (x + val).to_numpy()
+    expected = x.to_numpy() + val
+    np.testing.assert_equal(actual, expected, strict=True)
```

Comment:
`actual` and `expected` go through ndonnx. I think it would be better to test against NumPy directly:

```suggestion
@pytest.mark.parametrize(
    "val", [1, 1.0, 1.5, 0.123456789, "a", np.float64(0.123456789)]
)
def test_scalar_promote(val):
    promoted, = ndx._utility.promote(val)
    promoted_np = promoted.to_numpy()

    np.testing.assert_equal(promoted, promoted_np, strict=True)
```

Reply:
`actual`  goes through ndonnx, while in `expected` the addition (and hence type promotion) happens in NumPy.

Reply:
Oh sorry, the suggestion is very nonsensical! I meant:

```python
@pytest.mark.parametrize(
    "val", [1, 1.0, 1.5, 0.123456789, "a", np.float64(0.123456789)]
)
def test_scalar_promote(val):
    promoted, = ndx._utility.promote(val).to_numpy()
    np.testing.assert_equal(promoted, np.asarray(va), strict=True)
```

I.e. wouldn't it be nicer, in the spirit of unit testing, to explicitly call `promote` in the test? This would also remove any implicit assumptions to what may be happening to a `val` on its way through `ndx.asarray(val).to_numpy()`.

Reply:
Is that what we want though? Consider:

```python
(Pdb) np.asarray(1.).dtype
dtype('float64')
```

```python
(Pdb) (np.asarray(1., dtype=np.float32) + 1.).dtype
dtype('float32')
```

i.e. creation seems to go to the default int/float dtype (not the smallest dtype). Promotion with scalars is independent to this as far as I understand it.

Reply:
Ah, I see! Thanks! From a testing perspective I still think we should call `promote` directly, since that is ultimately what we want to unit test here:
```python
promoted, = ndx._utility.promote(val).to_numpy()
```
But then we should probably explicitly check that `promoted` is what we want it to be.

Reply:
Sure I don't mind either way to be honest. The previous way one could compare with NumPy.